packages/flutter_blue/tizen/src/bluetooth_device_controller.h

```diff
@@ -0,0 +1,103 @@
+#ifndef FLUTTER_BLUE_TIZEN_BLUETOOTH_DEVICE_CONTROLLER_H
+#define FLUTTER_BLUE_TIZEN_BLUETOOTH_DEVICE_CONTROLLER_H
+
+#include "GATT/bluetooth_service.h"
+#include "bluetooth.h"
+#include "flutterblue.pb.h"
+#include "notifications_handler.h"
+#include "utils.h"
+
+namespace flutter_blue_tizen {
+
+class BluetoothDeviceController {
+ public:
+  enum class State {
+    kConnected,
+    kConnecting,
+    kDisconnected,
+    kDisconnecting,
+  };
+
+  using requestMtuCallback =
+      std::function<void(bool, const BluetoothDeviceController&)>;
+
+  BluetoothDeviceController(
+      const std::string& address,
+      NotificationsHandler& notifications_handler) noexcept;
+
+  BluetoothDeviceController(
+      const char* address,
+      NotificationsHandler& notifications_handler) noexcept;
+
+  ~BluetoothDeviceController() noexcept;
+
+  BluetoothDeviceController() = delete;
+
+  BluetoothDeviceController(const BluetoothDeviceController& address) = delete;
+
+  const std::string& cAddress() const noexcept;
+
+  State GetState() const noexcept;
+
+  std::vector<proto::gen::BluetoothDevice>& ProtoBluetoothDevices() noexcept;
+
+  const std::vector<proto::gen::BluetoothDevice>& cProtoBluetoothDevices()
+      const noexcept;
+
+  void Connect(bool auto_connect);
+
+  void Disconnect();
+
+  static void ConnectionStateCallback(int result, bool connected,
+                                      const char* remote_address,
+                                      void* user_data) noexcept;
+
+  static bt_gatt_client_h GetGattClient(const std::string& address);
+
+  static void DestroyGattClientIfExists(const std::string& address) noexcept;
+
+  static proto::gen::DeviceStateResponse_BluetoothDeviceState
+  LocalToProtoDeviceState(const BluetoothDeviceController::State state);
+
+  void DiscoverServices();
+
+  std::vector<btGatt::PrimaryService*> GetServices() noexcept;
+
+  btGatt::PrimaryService* GetService(const std::string& uuid) noexcept;
+
+  uint32_t GetMtu() const;
+
+  void RequestMtu(uint32_t mtu, const requestMtuCallback& callback);
+
+  void NotifyDeviceState() const;
+
+  const NotificationsHandler& cNotificationsHandler() const noexcept;
+
+ private:
+  /**
+   * @brief all attributes are depentent on this mutex
+   */
+
+  std::mutex operation_mutex_;
+
+  std::vector<proto::gen::BluetoothDevice> proto_bluetooth_devices_;
```

Comment:
Why does `BluetoothDeviceController` manage multiple `BluetoothDevice`? Isn't this the role of a `BluetoothManager`?

Reply:
Depending on whether the `allow duplicates` option is on, this container would aggregate all the results from `scanCallback`. 
They all describe the same device.

Reply:
Does `BluetoothDeviceController` need to aggregate duplicated scan results? The only reason I see for having `BluetoothDeviceController.proto_bluetooth_devices_` is to later return some equal (not identical) proto::gen::BluetoothDevice objects in `BluetoothManager.GetConnectedProtoBluetoothDevices()`. I don't see why this API should return multiple "Connected Info" instances.

If you want to return scan result duplicates when `allow_duplicate` is true, `BluetoothManager.bluetooth_devices_` is effectively working as a cache of scanned devices. See https://github.com/boskokg/flutter_blue_plus/blob/259ebdb15a03e95a4f98280a7e7c4137f4171f88/android/src/main/java/com/boskokg/flutter_blue_plus/FlutterBluePlusPlugin.java#L833-L838.

Reply:
I believe that `RSSI` can differ.

Reply:
But the authors decided that they don't care as far as I can see in the code...

Reply:
Isn't rssi sent through `ScanResult` not `BluetoothDevice`? `BluetoothManager.GetConnectedProtoBluetoothDevices()` may return value-wise unique objects, while `BluetoothManager.ScanCallback()` may return `ScanResult`s with same bluetooth device info but with different rssi: https://github.com/HakkyuKim/plugins/commit/e7c7a3bff8cdf5edb1339c0bb5008c4aa50274dd